File: YuGiOh2/Cards/C70046172.cs
using System.Linq;
using YuGiOh2.Base;

namespace YuGiOh2.Cards
{
    /// <summary>
    /// 突进
    /// </summary>
    public class C70046172
    {
        public static int Type { get; } = (int)ChooseTargetType.AllFaceUpMonster;

        public static bool CheckIfAvailable(Player player)
        {
            return player.Field.MonsterFields.Any(c => c != null && !c.Status.FaceDown) ||
                player.Enemy.Field.MonsterFields.Any(c => c != null && !c.Status.FaceDown);
        }

        public static void ProcessEffect(string targetID, Player player)
        {
            if (targetID == null)
                return;

            MonsterCard target = player.Field.MonsterFields.FirstOrDefault(c => c != null && c.UID == targetID) ??
                player.Enemy.Field.MonsterFields.FirstOrDefault(c => c != null && c.UID == targetID);
            if (target == null)
                return;
            target.ATK += 700;
        }

        public static void ProcessEndPhase(string targetID, Player player)
        {
            if (targetID == null)
                return;

            MonsterCard target = player.Field.MonsterFields.FirstOrDefault(c => c != null && c.UID == targetID) ??
                player.Enemy.Field.MonsterFields.FirstOrDefault(c => c != null && c.UID == targetID);
            if (target == null)
                return;
            target.ATK = target.ATK > 700 ? target.ATK - 700 : 0;
        }
    }
}

File: YuGiOh2/Scripts/C70046172.lua
Type = 13;
--突进
function checkIfAvailable(player)
    for i, monster in pairs(player.Field.MonsterFields) do
        if monster ~= nil and not monster.Status.FaceDown then
            return true;
        end
    end
    for i, monster in pairs(player.Enemy.Field.MonsterFields) do
        if monster ~= nil and not monster.Status.FaceDown then
            return true;
        end
    end
    return false;
end

function processEffect(targetID, player)
    if targetID == nil then
        return ;
    end
    local target = nil;
    for i, monster in pairs(player.Field.MonsterFields) do
        if monster ~= nil and monster.UID == targetID then
            target = monster;
        end
    end
    for i, monster in pairs(player.Enemy.Field.MonsterFields) do
        if monster ~= nil and monster.UID == targetID then
            target = monster;
        end
    end
    if target == nil then
        return ;
    end
    target.ATK = target.ATK + 700;
end

function processEndPhase(targetID, player)
    if targetID == nil then
        return ;
    end
    local target = nil;
    for i, monster in pairs(player.Field.MonsterFields) do
        if monster ~= nil and monster.UID == targetID then
            target = monster;
        end
    end
    for i, monster in pairs(player.Enemy.Field.MonsterFields) do
        if monster ~= nil and monster.UID == targetID then
            target = monster;
        end
    end
    if target == nil then
        return ;
    end
    if target.ATK > 700 then
        target.ATK = target.ATK - 700;
    else
        target.ATK = 0;
    end
end